pkg/resource/clusteroperator.go

```diff
@@ -217,33 +218,57 @@ func isDeploymentStatusAvailableAndUpdated(deploy *appsapi.Deployment) bool {
 		deploy.Status.UpdatedReplicas == deploy.Status.Replicas
 }
 
-func (gco *generatorClusterOperator) syncVersions(op *configapi.ClusterOperator) (modified bool, err error) {
+// syncVersions updates versions reported through ClusterOperator resource.
+//
+// We report two distinct versions, one for the Operator and the other for the
+// Image Registry, we read the first from the environment and the latter from
+// an annotation on Registry's deployment.
+func (gco *generatorClusterOperator) syncVersions(op *configapi.ClusterOperator) (bool, error) {
+	var updated int
+
 	deploy, err := gco.deployLister.Get(imageregistryv1.ImageRegistryName)
-	if err != nil {
-		if kerrors.IsNotFound(err) {
-			err = nil
+	if err != nil && !kerrors.IsNotFound(err) {
+		// Return error if other than "NotFound".
+		return false, err
+	} else if err == nil {
+		v := deploy.Annotations[imageregistryv1.VersionAnnotation]
+		if len(v) > 0 && isDeploymentStatusAvailableAndUpdated(deploy) {
+			if gco.updateVersionInPlace(op, "image-registry", v) {
+				updated++
+			}
 		}
-		return
 	}
 
-	deploymentVersion := deploy.Annotations[imageregistryv1.VersionAnnotation]
-	if len(deploymentVersion) == 0 || !isDeploymentStatusAvailableAndUpdated(deploy) {
-		return
+	if v := os.Getenv("RELEASE_VERSION"); len(v) > 0 {
+		if gco.updateVersionInPlace(op, "operator", v) {
+			updated++
+		}
```

Comment:
@ricardomaraschini I think my own confusion around when to report the `operator` version led us astray here.

We need the following logic:

1. If the operator is `Managed`, we can't update the `operator` version until the image registry has rolled out (via `isDeploymentStatusAvailableAndUpdated`)
2. For all other management states, report the `operator` version (and truthfully we shouldn't report any other version number).